File: example/1d-linear-convection/weno3/python/04m/reconstructor/weno3.py
# reconstructor/weno3.py
import numpy as np
from .base import Reconstructor

class Weno3Reconstructor(Reconstructor):
    def reconstruct(self, q, cfd):
        domain = cfd.domain
        solution = cfd.solution
        self._reconstruct_left_interfaces(domain, q, solution.q_face_left)
        self._reconstruct_right_interfaces(domain, q, solution.q_face_right)
        
    def _reconstruct_left_interfaces(self, domain, u, qL):
        """在每个 i+1/2 界面，计算左单元贡献的 qL (即 u_{i+1/2}^-)"""
        for i in range(domain.ist - 1, domain.ied):
            j = i - (domain.ist - 1)
            v1, v2, v3 = u[i-1], u[i], u[i+1]
            qL[j] = self._reconstruct_from_right_biased_stencil(v3, v2, v1)
            
    def _reconstruct_right_interfaces(self, domain, u, qR):
        """在每个 i+1/2 界面，计算右单元贡献的 qR (即 u_{i+1/2}^+)"""
        for i in range(domain.ist, domain.ied + 1):
            j = i - domain.ist
            v1, v2, v3 = u[i-1], u[i], u[i+1]
            qR[j] = self._reconstruct_from_left_biased_stencil(v3, v2, v1)

    def _reconstruct_from_left_biased_stencil(self, v1, v2, v3):
        """使用左偏 stencil (v1,v2,v3) 重建界面值（对应 u_{i+1/2}^+）"""
        eps = 1e-6
        beta0 = (v3 - v2)**2      # smoothness indicator for stencil [v2, v3]
        beta1 = (v2 - v1)**2      # smoothness indicator for stencil [v1, v2]
        
        d0, d1 = 2/3, 1/3         # optimal linear weights (for right value)
        alpha0 = d0 / (eps + beta0)**2
        alpha1 = d1 / (eps + beta1)**2
        w0 = alpha0 / (alpha0 + alpha1)
        w1 = alpha1 / (alpha0 + alpha1)
        
        q0 = 1.5 * v2 - 0.5 * v3  # reconstruction from [v2, v3]
        q1 = 0.5 * v1 + 0.5 * v2  # reconstruction from [v1, v2]
        return w0 * q0 + w1 * q1

    def _reconstruct_from_right_biased_stencil(self, v1, v2, v3):
        """使用右偏 stencil (v1,v2,v3) 重建界面值（对应 u_{i+1/2}^-）"""
        eps = 1e-6
        beta0 = (v3 - v2)**2
        beta1 = (v2 - v1)**2
        
        d0, d1 = 1/3, 2/3         # optimal linear weights (for left value)
        alpha0 = d0 / (eps + beta0)**2
        alpha1 = d1 / (eps + beta1)**2
        w0 = alpha0 / (alpha0 + alpha1)
        w1 = alpha1 / (alpha0 + alpha1)
        
        q0 = 1.5 * v2 - 0.5 * v3  # from [v2, v3]
        q1 = 0.5 * v1 + 0.5 * v2 # from [v1, v2]
        return w0 * q0 + w1 * q1

File: example/1d-linear-convection/weno3/fortran/cfd/05d/src/reconstruct/weno3.f90
! ==================== weno3.f90 ================================
! WENO (Weighted Essentially Non-Oscillatory) reconstructor for OneFLOW-CFD
!
! Purpose: Implement WENO reconstruction scheme
! Reference: Jiang & Shu (1996)
! Author: OneFLOW-CFD Team
! Date: Created
! ===============================================================

module weno3_module
    use kinds, only: dp
    use domain_module, only: ComputationalDomainType
    use solution_module, only: SolutionType
    implicit none
    
    private
    
    ! WENO重构器类型
    type, public :: WenoReconstructorType
    contains
        procedure :: reconstruct => weno_reconstruct
    end type WenoReconstructorType
    
    ! WENO相关参数
    real(dp), parameter, public :: eps_weno = 1.0e-6_dp
    
contains
    
    ! WENO reconstruction
    subroutine weno_reconstruct(this, q, domain, solution)
        class(WenoReconstructorType), intent(inout) :: this
        real(dp), intent(in) :: q(:)
        type(ComputationalDomainType), intent(in) :: domain
        type(SolutionType), intent(inout) :: solution
        
        call weno3L_periodic(q, solution%q_face_left, domain, solution)
        call weno3R_periodic(q, solution%q_face_right, domain, solution)
    end subroutine weno_reconstruct
	
    ! WENO-3 reconstruction for left interface
    subroutine weno3L_periodic(u, qL, domain, solution)
        real(dp), intent(in) :: u(:)
        real(dp), intent(out) :: qL(:)
        type(ComputationalDomainType), intent(in) :: domain
        type(SolutionType), intent(inout) :: solution
        
        integer :: i, j, nghosts, ist, ied
        
        nghosts = domain%nghosts
        ist = domain%ist
        ied = domain%ied
        
        do i = ist-1, ied-1
            j = i - (ist - 1) + 1
            qL(j) = wc3L(u(i-1), u(i), u(i+1))
        end do
        
    end subroutine weno3L_periodic
    
    ! WENO-3 reconstruction for right interface
    subroutine weno3R_periodic(u, qR, domain, solution)
        real(dp), intent(in) :: u(:)
        real(dp), intent(out) :: qR(:)
        type(ComputationalDomainType), intent(in) :: domain
        type(SolutionType), intent(inout) :: solution
        
        integer :: i, j, nghosts, ist, ied
        
        nghosts = domain%nghosts
        ist = domain%ist
        ied = domain%ied
        
        do i = ist, ied
            j = i - ist + 1
            qR(j) = wc3R(u(i-1), u(i), u(i+1))
        end do
        
    end subroutine weno3R_periodic
    
    ! WENO-3 nonlinear weights for left-biased stencil
    function wc3L(v1, v2, v3) result(f)
        real(dp), intent(in) :: v1, v2, v3
        real(dp) :: f, s0, s1, d0, d1, c0, c1, w0, w1, q0, q1
        
        ! Smoothness indicators
        s0 = (v3 - v2)**2
        s1 = (v2 - v1)**2
        
        ! Nonlinear weights
        d0 = 2.0_dp/3.0_dp
        d1 = 1.0_dp/3.0_dp
        
        c0 = d0 / ((eps_weno + s0)**2)
        c1 = d1 / ((eps_weno + s1)**2)
        
        w0 = c0 / (c0 + c1)
        w1 = c1 / (c0 + c1)
        
        ! Candidate stencils
        q0 = 0.5_dp * v2 + 0.5_dp * v3
        q1 = -0.5_dp * v1 + 1.5_dp * v2
        
        ! Reconstructed value
        f = w0 * q0 + w1 * q1
    end function wc3L
    
    ! WENO-3 nonlinear weights for right-biased stencil
    function wc3R(v1, v2, v3) result(f)
        real(dp), intent(in) :: v1, v2, v3
        real(dp) :: f, s0, s1, d0, d1, c0, c1, w0, w1, q0, q1
        
        ! Smoothness indicators
        s0 = (v2 - v1)**2
        s1 = (v3 - v2)**2
        
        ! Nonlinear weights
        d0 = 2.0_dp/3.0_dp
        d1 = 1.0_dp/3.0_dp
        
        c0 = d0 / ((eps_weno + s0)**2)
        c1 = d1 / ((eps_weno + s1)**2)
        
        w0 = c0 / (c0 + c1)
        w1 = c1 / (c0 + c1)
        
        ! Candidate stencils
        q0 = 0.5_dp * v1 + 0.5_dp * v2
        q1 = 1.5_dp * v2 - 0.5_dp * v3
        
        ! Reconstructed value
        f = w0 * q0 + w1 * q1
    end function wc3R	
    
end module weno3_module